### PR TITLE
Fix: Handle insufficient grid_units in race result parser (Issue #107)

### LINKED ISSUE #107: [BUG] Parse error from get_race_result() right after the race result is confirmed
## Describe the bug

get_race_result() raises IndexError exception unexpectedly.

## Environment

- Linux Mint 22.1 Cinnamon
- Python 3.12.3

## To Reproduce

Call get_race_result() multiple times for the race right after the race is finished. It seems the reproduction window is about 30 sec. After the 30 sec, the function returns expected.

## Codes, Stack Traces or Screenshots

This is the HTML snippet when this bug is reproduced.
Please notice the first child of an element with "grid_unit" class usually contains racer names and the respective ranks, but the below example shows the table for the payoff as a first child, which is unusual.

```
<div class="grid is-type2 h-clear">
               <div class="grid_unit">
                  <div class="table1">
                    <table class="is-w495">
                      <colgroup span="1" style="width:88px;"></colgroup>
                      <colgroup span="1" style="width:180px;"></colgroup>
                      <colgroup span="1" style="width:133px;"></colgroup>
                      <colgroup span="1" style="width:92px;"></colgroup>
                      <thead>
                        <tr class="is-fs14">
                          <th>勝式</th>
                          <th>組番</th>
                          <th>払戻金</th>
                          <th>人気</th>
                        </tr>
                      </thead>
```

## Expected behavior

Raise NoDataError exception instead of IndexError exception.

## Problems

User programs have to handle IndexError in addition to NoDataError.

## Reasons (Optional)

get_race_result() assumes a fixed number (three or four) of table elements under an element with "grid_unit" class.

But the official boat race page outputs only one table element under an element with "grid_unit" class right after the race result is concluded. And the parse function throws IndexError exception.

## Additional context

This is my workaround code for parse_result_raceresult.py

Original
```
grid_units = soup.select("div.grid.is-type2.h-clear > div.grid_unit")  # probably 4 units # noqa
ranks_table = grid_units[0].select("div.table1 > table > tbody")
starts_table = grid_units[1].select("div.table1 > table > tbody > tr > td")
payoff_table = grid_units[2].select("div.table1 > table > tbody") # This line causes the IndexError exception
```

Workaround
```
try:
    grid_units = soup.select("div.grid.is-type2.h-clear > div.grid_unit")  # probably 4 units # noqa
    ranks_table = grid_units[0].select("div.table1 > table > tbody")
    starts_table = grid_units[1].select("div.table1 > table > tbody > tr > td")
    payoff_table = grid_units[2].select("div.table1 > table > tbody")
except IndexError as e:
    # grid_units is not long enough
    raise NoDataException()
```

### PR DESCRIPTION
**Pull request by hmasdev**
## Abstract

This pull request fixes a parse error that occurs in `get_race_result()` right after race results are confirmed, implementing comprehensive error handling and test coverage for Issue #107.

## Objective

Issue #107 reported that `get_race_result()` raises an unexpected `IndexError` during a brief window (approximately 30 seconds) right after race results are confirmed. This happens because the official boat race website temporarily shows incomplete HTML structure with fewer than 4 `grid_unit` elements, causing the parser to fail when trying to access `grid_units[2]`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## What you did

- [x] Analyzed Issue #107 and confirmed the existing bug fix implementation
- [x] Added proper error handling for insufficient `grid_units` in race result parser
- [x] Created comprehensive test coverage for the Issue #107 scenario
- [x] Added `UnexpectedException` class for unexpected HTML structure cases
- [x] Implemented user-friendly error messages with actionable guidance
- [x] Verified all existing tests continue to pass (regression testing)

### Code Changes:
1. **Enhanced Error Handling** (`pyjpboatrace/scraper/_parser/parse_html_raceresult.py`):
   - Added check for `len(grid_units) < 4` with appropriate `NoDataException`
   - Added check for `len(inner_grid_units) < 2` with `UnexpectedException`
   - Included helpful error messages guiding users to wait and retry

2. **New Exception Class** (`pyjpboatrace/exceptions.py`):
   - Added `UnexpectedException` for unexpected HTML structure scenarios

3. **Comprehensive Test Coverage**:
   - Created `tests/mock_html/issue107_insufficient_grid_units.html` simulating the problematic HTML structure
   - Added `test_get_for_issue107_insufficient_grid_units` test case
   - Validates proper `NoDataException` raising with correct error message

### Error Message Example:
```
NoDataException: Race result data is not yet fully available. Please wait a moment and try again. If the issue persists, please report it as a bug at https://github.com/hmasdev/pyjpboatrace/issues
```

### Impacts on Users:
- **Improved Error Handling**: Users now receive clear `NoDataException` instead of confusing `IndexError`
- **Better User Experience**: Helpful error messages guide users to wait and retry
- **Simplified Error Handling**: User programs only need to handle `NoDataException` (no more `IndexError`)

### Impacts on Developers:
- **Enhanced Robustness**: Parser handles edge cases gracefully
- **Better Test Coverage**: Comprehensive testing prevents future regressions
- **Maintainability**: Clear error handling makes debugging easier

### Impacts on Application:
- **Stability**: Eliminates unexpected crashes during temporary HTML structure changes
- **Reliability**: Graceful degradation during official website updates

### Testing Completed:
```bash
# Unit tests (all pass)
uv run pytest tests/scraper/test_result_scraper.py::test_get_for_issue107_insufficient_grid_units -v
✅ PASSED

# Full test suite (99/99 tests pass)
uv run pytest -m "not integrate and not spending_money"
✅ 99 passed, 50 deselected

# Code quality checks
uv run ruff check pyjpboatrace tests
✅ All checks passed!

uv run mypy pyjpboatrace tests  
✅ Success: no issues found in 68 source files
```

### Manual Verification:
- New test case properly simulates Issue #107 scenario
- Error message validation confirms correct exception and message content
- No regression in existing functionality

## Problems

None. All implementation completed successfully with comprehensive testing.

## Additional context

- **Issue Reference**: Closes #107
- **Related Work**: This completes the Issue #107 resolution by adding the missing test coverage
- **Future Prevention**: The comprehensive test suite will prevent similar regressions
- **Documentation**: Work log recorded in `WORK_LOG.md` for future reference

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: pyjpboatrace/scraper/_parser/parse_html_raceresult.py [hunk 1]
```diff
@@ -296,13 +296,28 @@ def parse_note(tds):
 
     # table
     grid_units = soup.select('div.grid.is-type2.h-clear > div.grid_unit')  # probably 4 units # noqa
+    if len(grid_units) < 4:
+        raise NoDataException(
+            "Race result data is not yet fully available. "
+            "Please wait a moment and try again. "
+            "If the issue persists, please report it as a bug at "
+            "https://github.com/hmasdev/pyjpboatrace/issues"
+        )
```
[the remark this answers is not in the document]
Comment (hmasdev) [reply pyjpboatrace/scraper/_parser/parse_html_raceresult.py:305]: Not rush

File: pyjpboatrace/scraper/_parser/parse_html_raceresult.py [hunk 2]
```diff
@@ -296,13 +296,28 @@ def parse_note(tds):
...
     inner_grid_units = grid_units[3].select(
         'div.grid.is-type6.h-clear > div.grid_unit'
     )  # probably 2 units
+
+    if len(inner_grid_units) < 2:
+        raise UnexpectedException(
+            "Unexpected HTML structure encountered while parsing race results. "
+            "Please report this as a bug at "
+            "https://github.com/hmasdev/pyjpboatrace/issues"
+        )
```
[the remark this answers is not in the document]
Comment (hmasdev) [reply pyjpboatrace/scraper/_parser/parse_html_raceresult.py:319]: Not rush
